### PR TITLE
feat(ivy): add support for typings in ngcc (and some minor refactorings)

### PR DESCRIPTION
**Pull request by gkalpak**
Sits on top of #25090 (so the first 7 commits are from there).

## PR Checklist
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
```
[x] Feature
[x] Refactoring (no functional changes, no api changes)
```

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

File: packages/compiler-cli/src/ngcc/src/transform/package_transformer.ts [hunk 1]
```diff
@@ -124,6 +137,36 @@ export class PackageTransformer {
...
+      analyzedFile.analyzedClasses.forEach(
+          analyzedClass =>
+              dtsTransformer.recordStaticField(analyzedClass.name, analyzedClass.compilation));
+
+      // Find the corresponding `.d.ts` file.
+      const sourceFileName = analyzedFile.sourceFile.fileName;
+      const originalDtsFileName = entryPoint.getDtsFileNameFor(sourceFileName);
+      const originalDtsContents = readFileSync(originalDtsFileName, 'utf8');
+
+      // Tranform the `.d.ts` file based on the recorded source file changes.
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngcc/src/transform/package_transformer.ts]: Tranform => Transform

**Comment by gkalpak** [comment]
Rebased, resolved conflicts, and addressed comment.

**Comment by petebacondarwin** [comment]
@gkalpak we have rolled this PR into the `ngcc-integration-test` branch and associated PR: https://github.com/angular/angular/pull/25406

We need to add any changes you make here to that, because @alxhub is going to try to land the whole lot in one go.

**Comment by gkalpak** [comment]
Rebased #25406 on top of this ✔️

**Comment by manughub** [comment]
Relates to FW-82

**Comment by petebacondarwin** [comment]
Superceded by #25406

**Pull request closed without merging**